### PR TITLE
Refactor vamp

### PR DESCRIPTION
**Pull request by clonker**
created a hierarchy of Koopman models so that also TICA can be integrated in a second PR. Fixed documentation where appropriate.
VAMP is now estimated by a covariance model, but has a static `from_data` method.

### CODE REVIEW

File: docs/source/notebooks/vamp.ipynb [hunk 1]
```diff
@@ -91,7 +90,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Given the data, all that is left to do is fitting a model and obtaining it from the estimator."
+    "Given the data, all that is left to do is fitting a model by providing covariances and obtaining it from the estimator or using the [from_data](../api/generated/sktime.decomposition.VAMP.rst#sktime.decomposition.VAMP.from_data) shortcut directly."
```
Comment (brookehus) [inline docs/source/notebooks/vamp.ipynb:93]: grammar "fit a model"

File: docs/source/notebooks/vamp.ipynb [hunk 2]
```diff
@@ -100,8 +99,15 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "vamp_estimator.fit(feature_trajectory)\n",
-    "model = vamp_estimator.fetch_model();"
+    "covars = sktime.decomposition.VAMP.covariance_estimator(\n",
+    "    lagtime=1  # the time lag under which the covariance matrices are estimated\n",
```
Comment (brookehus) [inline docs/source/notebooks/vamp.ipynb:103]: with which?

File: tests/decomposition/test_vamp.py [hunk 3]
```diff
@@ -128,10 +130,10 @@ def do_test(self, dim, rank, test_partial_fit=False):
 
             # vamp2.singular_values # trigger diagonalization
             for t, ref in zip(trajs, phi_trajs):
-                assert_allclose_ignore_phase(vamp2.transform(t[tau:], right=True), ref, rtol=rtol, atol=atol)
+                assert_allclose_ignore_phase(vamp2.transform(t[tau:], forward=False), ref, rtol=rtol, atol=atol)
 
             for t, ref in zip(trajs, psi_trajs):
-                assert_allclose_ignore_phase(vamp2.transform(t[0:-tau], right=False), ref, rtol=rtol, atol=atol)
+                assert_allclose_ignore_phase(vamp2.transform(t[0:-tau], forward=True), ref, rtol=rtol, atol=atol)
```
Comment (brookehus) [inline tests/decomposition/test_vamp.py:136]: def like forward better than right

File: tests/decomposition/test_vamp.py [hunk 4]
```diff
@@ -205,21 +208,19 @@ def test_singular_functions_against_MSM(self):
...
-        self.vamp.right = False
+        phi = self.vamp.transform(np.eye(3), forward=False)
         psi = self.vamp.transform(np.eye(3))
         assert_allclose_ignore_phase(U, psi, atol=1E-5)
         assert_allclose_ignore_phase(V, phi, atol=1E-5)
 
         cumsum_Tsym = np.cumsum(S[1:] ** 2)
         cumsum_Tsym /= cumsum_Tsym[-1]
-        np.testing.assert_allclose(self.vamp.cumvar, cumsum_Tsym)
+        np.testing.assert_allclose(self.vamp.cumulative_kinetic_variance, cumsum_Tsym)
```
Comment (brookehus) [inline tests/decomposition/test_vamp.py:218]: :) +1 for this change

File: tests/markov/msm/test_mlmsm.py [hunk 5]
```diff
@@ -374,3 +374,12 @@ def test_raises_disconnected(disconnected_states, lag, count_mode):
...
         MaximumLikelihoodMSM(reversible=True).fit(submodel)
+
+
+def test_nonreversible_disconnected():
+    msm1 = MarkovStateModel([[.7, .3], [.3, .7]])
+    msm2 = MarkovStateModel([[.9, .05, .05], [.3, .6, .1], [.1, .1, .8]])
+    traj = np.concatenate([msm1.simulate(10000), 2 + msm2.simulate(10000)])
+    counts = TransitionCountEstimator(lagtime=1, count_mode="sliding").fit(traj).fetch_model()
+    msm = MaximumLikelihoodMSM(reversible=False).fit(counts).fetch_model()
+    print(msm.transition_matrix)
```
Comment (brookehus) [inline tests/markov/msm/test_mlmsm.py:385]: is this test supposed to end in a print statement?

File: sktime/decomposition/koopman.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+
+    def __init__(self, operator: np.ndarray, basis_transform_forward: Optional[KoopmanBasisTransform],
+                 basis_transform_backward: Optional[KoopmanBasisTransform], cov: CovarianceModel,
+                 rank_0: int, rank_t: int, dim=None, scaling=None, epsilon=1e-6):
+        r""" For a description of parameters `operator`, `basis_transform_forward`, `basis_transform_backward`,
+        and `output_dimension`: please see :meth:`KoopmanModel.__init__`.
+
+        Parameters
+        ----------
+        cov : CovarianceModel
```
Comment (brookehus) [inline sktime/decomposition/koopman.py:254]: `cov` is a model?

File: sktime/decomposition/koopman.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+        r""" Rank of the time-lagged whitening transformation :math:`C_{tt}^{-1/2}`. """
+        return self._rank_t
+
+    @property
+    def epsilon(self) -> float:
+        r""" Singular value cutoff. """
+        return self._epsilon
+
+    @staticmethod
+    def _cumvar(singular_values) -> np.ndarray:
```
Comment (brookehus) [inline sktime/decomposition/koopman.py:345]: better name?

File: sktime/decomposition/koopman.py [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+                             "can only be in the interval (0, 1].")
+        elif value is not None and not isinstance(value, (Integral, Real)):
+            raise ValueError("Invalid type for dimension, got {}".format(value))
+        self._dim = value
+        self._output_dimension = CovarianceKoopmanModel.effective_output_dimension(
+            self.whitening_rank_0, self.whitening_rank_t, self._dim, self.singular_values
+        )
+
+    @property
+    def cumulative_kinetic_variance(self) -> np.ndarray:
```
Comment (brookehus) [inline sktime/decomposition/koopman.py:392]: how does this, or does it not, relate to `_cumvar`

File: sktime/decomposition/koopman.py [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+        if test_model is None:
+            test_model = self
+        Uk = self.singular_vectors_left[:, 0:self.output_dimension]
+        Vk = self.singular_vectors_right[:, 0:self.output_dimension]
+        res = None
+        if score_method == 'VAMP1' or score_method == 'VAMP2':
+            A = spd_inv_sqrt(Uk.T.dot(test_model.cov_00).dot(Uk), epsilon=self.epsilon)
+            B = Uk.T.dot(test_model.cov_0t).dot(Vk)
+            C = spd_inv_sqrt(Vk.T.dot(test_model.cov_tt).dot(Vk), epsilon=self.epsilon)
+            ABC = mdot(A, B, C)
```
Comment (brookehus) [inline sktime/decomposition/koopman.py:448]: i've seen this ABC notation before eg in our paper w/ martin but maybe adding a bit to the docs w/ the equation would be nice - sorry if you did and i missed it

File: sktime/decomposition/koopman.py [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+
+        statistics_mean_free : bool, default=False
+            If true, coefficients in `statistics` refer to the input
+            features with feature means removed.
+            If false, coefficients in `statistics` refer to the
+            unmodified input features.
+
+        Notes
+        -----
+        A "future expectation" of a observable g is the average of g computed
```
Comment (brookehus) [inline sktime/decomposition/koopman.py:498]: should this be connected to the forward/backward paradigm

File: sktime/decomposition/koopman.py (hunk 6 above)
Comment (clonker) [reply sktime/decomposition/koopman.py:254]: yes on both accounts, cov is a model containing `cov_00`, `cov_0t`, `cov_tt` and instantaneous as well as time-lagged means

File: sktime/decomposition/koopman.py (hunk 8 above)
Comment (clonker) [reply sktime/decomposition/koopman.py:392]: `_cumvar` is a static (ie stateless) private method which gets called by this property, agreed the name is perhaps not the best. on the other hand "cum" in the sense of cumulative is fairly standard (see `np.cumsum`) and so is `var` as a form of variance

File: sktime/decomposition/koopman.py (hunk 9 above)
Comment (clonker) [reply sktime/decomposition/koopman.py:448]: i can document this better, it is c&p code

File: sktime/decomposition/koopman.py (hunk 10 above)
Comment (clonker) [reply sktime/decomposition/koopman.py:498]: definitely connected but also only valid in the vamp/tica case id imagine

File: tests/markov/msm/test_mlmsm.py (hunk 5 above)
Comment (clonker) [reply tests/markov/msm/test_mlmsm.py:385]: you can ignore this for now, stuff for the next pr - sorry that a little bit of WIP bled into this changeset

**Pull request merged**